sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DiagnosticsProvider.java

```diff
@@ -547,11 +578,32 @@ public <T> Mono<CosmosItemResponse<T>> traceEnabledCosmosItemResponsePublisher(
      * @param publisher publisher to run.
      * @return wrapped publisher.
      */
-    public <T> Flux<T> runUnderSpanInContext(Flux<T> publisher) {
+    public <T> Flux<T> runUnderSpanInContext(Flux<T> publisher, CosmosPagedFluxOptions options) {
+
+        final double samplingRateSnapshot = clientTelemetryConfigAccessor.getSamplingRate(this.telemetryConfig);
+
+        options.setSamplingRateSnapshot(samplingRateSnapshot);
+
+        if (shouldSampleOutOperation(samplingRateSnapshot)) {
+            return publisher;
+        }
+
         return propagatingFlux
             .flatMap(ignored -> publisher);
     }
 
+    private boolean shouldSampleOutOperation(double samplingRate) {
+        if (samplingRate == 1) {
+            return false;
+        }
+
+        if (samplingRate == 0) {
+            return true;
+        }
+
+        return ThreadLocalRandom.current().nextDouble() >= samplingRate;
```

Comment:
why this random check? don't we need to sample whenever it is less than 1 and greater than equal to 0?

Also, why not use a random number generator? AFAIK, `ThreadLocalRandom` is used for concurrency, and here we don't need it. Might have an extra overhead.

Reply:
Random isn't thread safe
So, you would have to sync on it because this is used multi-threaded

Reply:
hmm, my understanding is Random will cause contention between threads, as it is thread safe. ThreadLocalRandom has less contention

Reply:
https://docs.oracle.com/javase/8/docs/api/java/util/Random.html